### PR TITLE
Reset nativeMap to null in onDestroy

### PR DESCRIPTION
**Pull request by Chaoba**
### Summary of changes
In `MapboxMap` and `Style`, the reference to `nativeMap` will be reset to null in onDestroy, so any method calls related to `nativeMap` will throw `MapboxMapMemoryLeakException`

`tearDown` method in BaseMapTest.kt is removed to avoid duplicate method call to `MapboxMap.onDestroy()` as it will be triggerd by lifecycle plugin.

Automatically unsubscribe all observers in MapboxMap when `MapboxMap.onDestroy()` is invoked.

## Pull request checklist:
 - [x] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
    
 - [ ] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Optimize code for java consumption (`@JvmOverloads`, `@file:JvmName`, etc).
 - [ ] Add example if relevant.
 - [ ] Document any changes to public APIs.
 - [ ] Run `make update-api` to update generated api files, if there's public API changes, otherwise the `verify-kotlin-binary-compatibility` CI will fail.
 - [x] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [ ] Add an entry inside this element for inclusion in the `mapbox-maps-android` changelog: `<changelog></changelog>`.
 - [x] If this PR is a `v10.[version]` release branch fix / enhancement, merge it to `main` firstly and then port to `v10.[version]` release branch.

Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).

### CODE REVIEW

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt [hunk 1]
```diff
@@ -394,15 +396,15 @@ abstract class AnnotationManagerImpl<G : Geometry, T : Annotation<G>, S : Annota
   }
 
   private fun updateDragSource() {
-    delegateProvider.getStyle { style ->
+    style.get()?.let { style ->
```
Comment (yunikkk) [inline plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:399]: What if style of MapboxMap changes? Will the AnnotationManager use outdated style?

Comment (Chaoba) [reply plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:399]: There is no need to change annotation manager, I'm working on a new solution.

File: app/src/androidTest/java/com/mapbox/maps/testapp/BaseMapTest.kt [hunk 2]
```diff
@@ -38,21 +37,6 @@ abstract class BaseMapTest {
...
-        mapView.onStop()
-        mapView.onDestroy()
-        latch.countDown()
-      }
-    }
-    if (!latch.await(10000, TimeUnit.MILLISECONDS)) {
-      throw TimeoutException()
-    }
-  }
-
```
Comment (tobrun) [inline app/src/androidTest/java/com/mapbox/maps/testapp/BaseMapTest.kt:55]: Could you explain the reason for this change, changes like these are great to highlight in your PR description.

File: app/src/main/java/com/mapbox/maps/testapp/examples/DebugModeActivity.kt [hunk 3]
```diff
@@ -172,12 +171,6 @@ class DebugModeActivity : AppCompatActivity() {
     }
   }
 
-  override fun onDestroy() {
-    super.onDestroy()
-    mapboxMap.unsubscribe(observable)
-    mapboxMap.unsubscribeResourceRequest(extensionObservable)
-  }
-
```
Comment (tobrun) [inline app/src/main/java/com/mapbox/maps/testapp/examples/DebugModeActivity.kt:180]: Could you explain the reason for this change, changes like these are great to highlight in your PR description.

File: sdk/src/test/java/com/mapbox/maps/MapboxMapTest.kt [hunk 4]
```diff
@@ -47,6 +48,10 @@ class MapboxMapTest {
     mapboxMap = MapboxMap(nativeMap, nativeObserver, styleObserver)
   }
 
+  @After
+  fun cleanUp() {
+    unmockkObject(MapProjectionUtils)
+  }
```
Comment (tobrun) [inline sdk/src/test/java/com/mapbox/maps/MapboxMapTest.kt:54]: nit add blank line below

**Review by tobrun — APPROVED** [review]
Thank you for adding tests!

File: app/src/androidTest/java/com/mapbox/maps/testapp/annotation/UpdateAnnotationTest.kt [hunk 5]
```diff
@@ -32,13 +32,16 @@ import java.util.concurrent.TimeoutException
 class UpdateAnnotationTest : BaseMapTest() {
   private val updateDelay = 100L
   private var index = 0
+  private var stopUpdate = false
```
Comment (ank27) [inline app/src/androidTest/java/com/mapbox/maps/testapp/annotation/UpdateAnnotationTest.kt:35]: is this related to reset `nativeMap` to null ?

File: plugin-annotation/src/test/java/com/mapbox/maps/plugin/annotation/AnnotationPluginImplTest.kt [hunk 6]
```diff
@@ -65,8 +65,9 @@ class AnnotationPluginImplTest {
 
     annotationPluginImpl.removeAnnotationManager(circleAnnotationManager)
     annotationPluginImpl.removeAnnotationManager(pointAnnotationManager)
-    annotationPluginImpl.removeAnnotationManager(polygonAnnotationManager)
-    annotationPluginImpl.removeAnnotationManager(polylineAnnotationManager)
+    assertEquals(2, annotationPluginImpl.managerList.size)
+
```
Comment (ank27) [inline plugin-annotation/src/test/java/com/mapbox/maps/plugin/annotation/AnnotationPluginImplTest.kt:69]: nit : empty line

File: sdk/src/androidTest/java/com/mapbox/maps/OfflineTest.kt [hunk 7]
```diff
@@ -545,7 +545,9 @@ class OfflineTest {
           .acceptExpired(true)
           .networkRestriction(NetworkRestriction.NONE)
           .build(),
-        {}
+        { progress ->
+          Logger.i("TAG", "completedResourceCount: ${progress.completedResourceCount}, requiredResourceCount: ${progress.requiredResourceCount}")
```
Comment (ank27) [inline sdk/src/androidTest/java/com/mapbox/maps/OfflineTest.kt:549]: nit : `TAG` instead of "TAG"

File: app/src/androidTest/java/com/mapbox/maps/testapp/BaseMapTest.kt (hunk 2 above)
Comment (Chaoba) [reply app/src/androidTest/java/com/mapbox/maps/testapp/BaseMapTest.kt:55]: Yeah, with the lifecycle plugin, `mapView.onDestroy()`  will be invoked automatically, if we keep this part it will be invoked twice.

File: app/src/androidTest/java/com/mapbox/maps/testapp/annotation/UpdateAnnotationTest.kt (hunk 5 above)
Comment (Chaoba) [reply app/src/androidTest/java/com/mapbox/maps/testapp/annotation/UpdateAnnotationTest.kt:35]: Yes, we need to stop the runnable after activity is destroyed, otherwise will trigger `MapboxMapMemoryLeakException`

File: app/src/main/java/com/mapbox/maps/testapp/examples/DebugModeActivity.kt (hunk 3 above)
Comment (Chaoba) [reply app/src/main/java/com/mapbox/maps/testapp/examples/DebugModeActivity.kt:180]: Observable will be automatically unsubscribed in `MapboxMap.onDestroy()`

File: plugin-annotation/src/test/java/com/mapbox/maps/plugin/annotation/AnnotationPluginImplTest.kt (hunk 6 above)
Comment (Chaoba) [reply plugin-annotation/src/test/java/com/mapbox/maps/plugin/annotation/AnnotationPluginImplTest.kt:69]: This is intentional as this block to verify 2 annotationManagers are removed

File: sdk/src/androidTest/java/com/mapbox/maps/OfflineTest.kt (hunk 7 above)
Comment (Chaoba) [reply sdk/src/androidTest/java/com/mapbox/maps/OfflineTest.kt:549]: Sharp eyes 😄

File: CHANGELOG.md [hunk 8]
```diff
@@ -5,7 +5,7 @@ Mapbox welcomes participation and contributions from everyone.
 # main
 ## Features ✨ and improvements 🏁
 * Optimize how plugins handle settings changes. Call `applySettings` only when settings value changes. ([#1189](https://github.com/mapbox/mapbox-maps-android/pull/1189))
-* Add new Exception type `MapboxMapMemoryLeakException` which will be thrown when there is a leak for nativeMap. ([1193](https://github.com/mapbox/mapbox-maps-android/pull/1193))
+* Add exception type `MapboxMapMemoryLeakException` which will be thrown when users try to invoke functions related with `nativeMap` in `MapboxMap` and `Style`. Automatically unsubscribe all observers in MapboxMap when `MapboxMap.onDestroy()` is invoked. ([1193](https://github.com/mapbox/mapbox-maps-android/pull/1193)) ([1202](https://github.com/mapbox/mapbox-maps-android/pull/1202))
```
Comment (yunikkk) [inline CHANGELOG.md:8]: Not sure how the wording should look like but current one seems super confusing

File: app/src/androidTest/java/com/mapbox/maps/testapp/annotation/UpdateAnnotationTest.kt (hunk 5 above)
Comment (yunikkk) [reply app/src/androidTest/java/com/mapbox/maps/testapp/annotation/UpdateAnnotationTest.kt:35]: I'm really worried that users might also face such cases in their apps. Do we really want to throw exceptions? Could we just print some error message and ignore the call instead of throwing?

File: sdk/src/main/java/com/mapbox/maps/MapboxMap.kt [hunk 9]
```diff
@@ -58,7 +59,7 @@ class MapboxMap :
     if (nativeMap == null) {
       throw MapboxMapMemoryLeakException()
     }
-    return nativeMap
+    return nativeMap!!
```
Comment (pengdev) [inline sdk/src/main/java/com/mapbox/maps/MapboxMap.kt:62]: nit: could be simplified to
`return nativeMap ?: throw MapboxMapMemoryLeakException()`

File: sdk/src/main/java/com/mapbox/maps/Style.kt [hunk 10]
```diff
@@ -20,15 +20,19 @@ import java.nio.ByteBuffer
...
+    styleManager = null
+  }
+
   private fun getStyleManager(): StyleManagerInterface {
-    if (styleManagerRef == null) {
+    if (styleManager == null) {
       throw MapboxMapMemoryLeakException()
     }
-    return styleManagerRef
+    return styleManager!!
```
Comment (pengdev) [inline sdk/src/main/java/com/mapbox/maps/Style.kt:35]: nit:
`return styleManager ?: throw MapboxMapMemoryLeakException()`

File: sdk/src/main/java/com/mapbox/maps/Style.kt [hunk 11]
```diff
@@ -20,15 +20,19 @@ import java.nio.ByteBuffer
...
  */
 class Style internal constructor(
-  private val styleManagerRef: StyleManagerInterface,
+  private var styleManager: StyleManagerInterface?,
   override val pixelRatio: Float
 ) : StyleInterface {
 
+  internal fun onDestroy() {
+    styleManager = null
+  }
```
Comment (pengdev) [inline sdk/src/main/java/com/mapbox/maps/Style.kt:29]: When is the style destroyed? seems currently it's destroyed when the MapboxMap is destroyed. I think to make the style object's lifecycle separate from MapboxMap's lifecycle, we could probably destroy the previous style when a new style is loaded? But I think it would be an behavioural breaking change, and many users might keep a reference to the style object for reuse, it could break their code.

File: CHANGELOG.md (hunk 8 above)
Comment (Chaoba) [reply CHANGELOG.md:8]: Will update and reword this part.

File: app/src/androidTest/java/com/mapbox/maps/testapp/annotation/UpdateAnnotationTest.kt (hunk 5 above)
Comment (Chaoba) [reply app/src/androidTest/java/com/mapbox/maps/testapp/annotation/UpdateAnnotationTest.kt:35]: Keep a reference to mapboxMap and Style actually and invoke methods after destroy is wrong behavior and may cause MML issue.  We throw exceptions here to let users fix this issue.

File: sdk/src/main/java/com/mapbox/maps/Style.kt (hunk 11 above)
Comment (Chaoba) [reply sdk/src/main/java/com/mapbox/maps/Style.kt:29]: Yes, we need to destroy the previous style when a new style is loaded.

**Review by pengdev — APPROVED** [review]
LGTM, imo we should differentiate the lifecycle of `MapboxMap` and `Style` object, and destroy the old style instance when new style is loaded, but it might cause behaviour breaking change if users are currently holding strong reference to the Style object for reuse in their application, we should revisit it in the next major release. I'm 👍  with current approach as an incremental improvement.
